File: top_interview_questions/2_medium/2_linked_list/odd_even_linked_list/solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head

        even = ListNode(head.val)
        odd = ListNode(head.next.val)
        even_head = even
        odd_head = odd
        head = head.next.next

        i = 0
        while head:
            if i % 2 == 0:
                even.next = ListNode(head.val)
                even = even.next
            else:
                odd.next = ListNode(head.val)
                odd = odd.next

            head = head.next
            i += 1

        even.next = odd_head
        return even_head

File: top_interview_questions/2_medium/2_linked_list/odd_even_linked_list/solution.go
package main

func main() {

}

type ListNode struct {
	Val  int
	Next *ListNode
}

func oddEvenList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	odd := &ListNode{Val: head.Val}
	even := &ListNode{Val: head.Next.Val}
	oddHead := odd
	evenHead := even

	head = head.Next.Next

	i := 0
	for head != nil {
		if i%2 == 0 {
			odd.Next = &ListNode{Val: head.Val}
			odd = odd.Next
		} else {
			even.Next = &ListNode{Val: head.Val}
			even = even.Next
		}

		head = head.Next
		i++
	}

	odd.Next = evenHead
	return oddHead
}
